src/com/mendix/widget/PullToRefresh/handlers/PullToRefresh.ts

```diff
@@ -0,0 +1,215 @@
+import * as domClass from "dojo/dom-class";
+import * as domStyle from "dojo/dom-style";
+
+interface Settings {
+    thresholdDistance: number;
+    maximumDistance: number;
+    reloadDistance: number;
+    mainElement: HTMLElement;
+    triggerElement: HTMLElement;
+    pullToRefreshElement: HTMLElement;
+    classPrefix: string;
+    cssProp: string;
+    iconArrow: string;
+    iconRefreshing: string;
+    pullToRefreshText: string;
+    releaseToRefreshText: string;
+    refreshText: string;
+    refreshTimeout: number;
+    onRefresh: ((value: Function) => Promise<void>) | ((value: Function) => void);
+    resistanceFunction: (value: number) => number;
+}
+
+export class PullToRefresh {
+    private settings: Settings;
+    private pullStart: {
+        screenX: number;
+        screenY: number;
+    };
+    private pullMoveY: number;
+    private distance = 0;
+    private distanceResisted = 0;
+    private state: "pending" | "pulling" | "releasing" | "refreshing" | "scrolling";
+    private setup: boolean;
+    private enable: boolean;
+    private timeout: number;
+    private updating: boolean;
+    private lastDistance = 0;
+
+    constructor(params: { pullToRefreshElement: HTMLElement, mainElement: HTMLElement }) {
+        this.onTouchStart = this.onTouchStart.bind(this);
+        this.onTouchMove = this.onTouchMove.bind(this);
+        this.onTouchEnd = this.onTouchEnd.bind(this);
+        this.onReset = this.onReset.bind(this);
+
+        this.settings = {
+            classPrefix: "pull-to-refresh-",
+            cssProp: "min-height",
+            iconArrow: "&#8675;",
+            iconRefreshing: "&hellip;",
+            mainElement: params.mainElement,
+            maximumDistance: 150,
+            onRefresh: () => { mx.ui.reload(); },
+            pullToRefreshElement: params.pullToRefreshElement,
+            pullToRefreshText: "Pull to refresh",
+            refreshText: "Refreshing",
+            refreshTimeout: 500,
+            releaseToRefreshText: "Release to refresh",
+            reloadDistance: 50,
+            resistanceFunction: (distance) => Math.min(1, distance / 2.5),
+            thresholdDistance: 80,
+            triggerElement: params.mainElement
+        };
+        this.state = "pending";
+        this.pullStart = { screenX: 0, screenY: 0 };
+    }
+
+    setupEvents() {
+        window.addEventListener("touchstart", this.onTouchStart);
+        window.addEventListener("touchend", this.onTouchEnd);
+        (window.addEventListener as WhatWGAddEventListener)("touchmove", this.onTouchMove, { passive: false });
+        this.setup = true;
+    }
+
+    removeEvents() {
+        window.removeEventListener("touchstart", this.onTouchStart);
+        window.removeEventListener("touchend", this.onTouchEnd);
+        (window.removeEventListener as WhatWGAddEventListener)("touchmove", this.onTouchMove, { passive: false });
+        this.setup = false;
+    }
+
+    private update() {
```

Comment:
Maybe this function should have a parameter `nextState` an `position` so it can handel all dom manupulations